tests/unit/s2n_client_server_name_extension_test.c

```diff
@@ -0,0 +1,169 @@
+    }
+
+    /* send */
+    {
+        struct s2n_connection *conn;
+        EXPECT_NOT_NULL(conn = s2n_connection_new(S2N_CLIENT));
+        EXPECT_SUCCESS(s2n_set_server_name(conn, test_server_name));
+
+        struct s2n_stuffer stuffer;
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&stuffer, 0));
+
+        EXPECT_SUCCESS(s2n_client_server_name_extension.send(conn, &stuffer));
+
+        uint16_t server_name_list_size;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint16(&stuffer, &server_name_list_size));
+        EXPECT_EQUAL(server_name_list_size, s2n_stuffer_data_available(&stuffer));
+
+        uint8_t name_type;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint8(&stuffer, &name_type));
+        EXPECT_EQUAL(name_type, 0);
+
+        uint16_t server_name_size;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint16(&stuffer, &server_name_size));
+        EXPECT_EQUAL(server_name_size, s2n_stuffer_data_available(&stuffer));
+        EXPECT_EQUAL(server_name_size, strlen(test_server_name));
+
+        char *server_name_data;
+        EXPECT_NOT_NULL(server_name_data = s2n_stuffer_raw_read(&stuffer, server_name_size));
+        EXPECT_BYTEARRAY_EQUAL(server_name_data, test_server_name, strlen(test_server_name));
+
+        EXPECT_SUCCESS(s2n_stuffer_free(&stuffer));
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* recv - basic */
+    {
+        struct s2n_connection *client_conn;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+        struct s2n_connection *server_conn;
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        struct s2n_stuffer stuffer;
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&stuffer, 0));
+
+        EXPECT_SUCCESS(s2n_set_server_name(client_conn, test_server_name));
+        EXPECT_SUCCESS(s2n_client_server_name_extension.send(client_conn, &stuffer));
+
+        EXPECT_STRING_EQUAL(server_conn->server_name, "");
+        EXPECT_SUCCESS(s2n_client_server_name_extension.recv(server_conn, &stuffer));
+        EXPECT_STRING_EQUAL(server_conn->server_name, test_server_name);
+
+        EXPECT_SUCCESS(s2n_stuffer_free(&stuffer));
+        EXPECT_SUCCESS(s2n_connection_free(client_conn));
+        EXPECT_SUCCESS(s2n_connection_free(server_conn));
+    }
+
+    /* recv - extra data ignored */
+    {
+        struct s2n_connection *client_conn;
+        EXPECT_NOT_NULL(client_conn = s2n_connection_new(S2N_CLIENT));
+
+        struct s2n_connection *server_conn;
+        EXPECT_NOT_NULL(server_conn = s2n_connection_new(S2N_SERVER));
+
+        struct s2n_stuffer stuffer;
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&stuffer, 0));
+
+        EXPECT_SUCCESS(s2n_set_server_name(client_conn, test_server_name));
+        EXPECT_SUCCESS(s2n_client_server_name_extension.send(client_conn, &stuffer));
+        EXPECT_SUCCESS(s2n_stuffer_write_uint8(&stuffer, 'a'));
+
+        EXPECT_STRING_EQUAL(server_conn->server_name, "");
```

Comment:
What happens when server_conn->server_name is already set? Do we overwrite the value?

Reply:
We actually do nothing. I can add a test for that.